### ISSUE TITLE
make partitioning of metrics more flexible

### ISSUE THREAD

**Issue by woodsaj** [raintank/tsdb-gw#16]
Now that metrictank supports clustering, we need to allow tsdb-gw to be more flexible with how we distribute metrics between kafka partitions.

Currently, we are generating an 8 byte key comprised of just the metric's OrgId.  This key is then hashed to determine the partition. This obviously won't be sufficient for single tenant deployments.

We need a more flexible partitioning scheme that can support numerous options. Though, right now we only really need to support partitioning byOrg or bySeries.

**Comment by woodsaj** [comment]
I guess the best approach here is to provide an Interface for for generating the []byte "key" used by sarama.  

Though this is a little complicated as we want to be able to generate the key for both schema.MetricData and schema.MetricDefinition

I have opened https://github.com/raintank/schema/issues/4  to extend schema package to provide a generic schema.Metric interface that both MetricData and MetricDefinition implement.

**Comment by Dieterbe** [comment]
makes sense

**Comment by Dieterbe** [comment]
So what problem do we really need to solve? I think:

*"the problem of accessing various internal MetricData and MetricDefinition attributes so we can generate partition keys based on them, while reusing the partition key generation code between MetricData and MetricDefinition"*

right?

If this code goes in the `schema` package, then it is pretty trivial to do this. Both types could expose the same `KeyByOrgId()`, `KeyBySeries()` etc functions that could internally reuse/call shared code. (or the code might be trivial enough to not even warrant reusing, eg `KeyBySeries` might just be a few lines)
Since the concept of a partition is already introduced in the `schema` package, it seems reasonable to just add that code there.  

If we want this code outside of the `schema` package, it could go in the metrictank package (for example), but note that both metrictank and tsdb-gw need to import it.
Also, then I think the concept of a partitioning should be moved out of `schema` and metrictank should have its own definition type that adds a partition field. (why should the partition field live in `schema`, but not code to set it?)

It looks like this is the approach you want to take.  While in general the interface idea seems like a good one, I don't think the proposed interface is a good solution here because it just presents the lowlevel attributes via a highlevel abstraction. the two don't jive well together IMHO.
Here's an alternative approach which I like more, and I think is also an established solution to the problem.  The problem now essentially becomes how to add methods to non-local types, which go does not allow. but the common solution seems to be declaring a type alias and declaring the methods (KeyByOrgId() etc) on that.  It has access to all attributes so it doesn't need to abstract them.  There's also no interface conversion overhead. Whenever you want to invoke the local methods (e.g. to set the partition) you would cast to the local type.

But personally I think if we're introducing the partition concept to the types in schema already, we may as well just add the code there too.

**Comment by woodsaj** [comment]
The partition number is a property of a metricDefinition and so belongs in the schema package.  But how that partition determined is external.

When submitting metrics via Kafka/tsdb-gw the partition is calculated by the sarama package.  But you could also use the carbon interface with metrictank and partition the metrics however you liked (using carbon-relay-ng or similar).  For example, users might want to partition on the Interval field so that high resolution metrics go to 1 metrictank and low resolution go to a different one.

I am not sure how your alias approach would work.  You cant have an alias that can represent 2 different struct types.  As far as i am aware, interfaces are the correct and only way to achieve what we need here.

**Comment by Dieterbe** [comment]
interface methods are meant to abstract away implementation details of some logic (so the Validate and SetId part of the interface looks good); not as a means to simply access attributes (and btw, the attributes are already public)

So instead of adding interface methods for something as low level as just getting the value of an attribute
you would just declare an alias (for both types). something like:
```
type MetricDefinitionSharded MetricDefinition

func (md MetricDefinitionSharded) KeyByOrg () []byte {
code1
}
func (md MetricDefinitionSharded) KeyBySeries () []byte {
code2
}

type MetricDataSharded MetricData

func (md MetricDataSharded) KeyByOrg () []byte {
code3
}
func (md MetricDataSharded) KeyBySeries () []byte {
code4
}
```

code1, code2, code3 and code4 are just a few lines, but if a need arises to have them reuse larger pieces of code, then they can just call another shared function.

**Comment by Dieterbe** [comment]
BTW I have nothing against interfaces by themselves, something like this would make a fine interface I think:
```
type Metric interface {
	Validate() error
	SetId()
        KeyByOrgId() []byte
        KeyBySeries() []byte
```

I really think we should just put the logic to generate the key in the schema package.
The types in the schema package come with their own code for serialization, so why couldn't it come with code to generate a key from their data?

**Comment by woodsaj** [comment]
I think this a good compromise.

My concern with this approach is that it is not extensible. The only options for partitioning are what we support in the schema package.

My intent was to allow easier development of future partitioning schemes.  But that is probably unnecessary at this stage. Adding another partitioning scheme would require code changes to metric tank anyway,  so there is little harm in also requiring changes to the schema package if needed.

**Comment by Dieterbe** [comment]
fixed by #20 , https://github.com/raintank/schema/pull/5 , https://github.com/raintank/schema/pull/6, https://github.com/raintank/schema/pull/7
